katdal/concatdata.py

```diff
@@ -200,44 +201,78 @@ def _initial_dtype(self):
 # -------------------------------------------------------------------------------------------------
 
 
+def infer_dtypes(sensor_data_sequence):
+    """List all known dtypes found in a sequence of sensor data objects.
+
+    This extracts the dtypes from a sequence of sensor data objects and
+    finds the unique set (where string types of various lengths are represented
+    by the string type with the maximum length). If all the objects have
+    unknown dtype, return [None] instead.
+
+    XXX The list of dtypes can be further collapsed to a single dtype if needed
+    via NumPy coercion. This will effectively forgo any relative type checking.
+
+    Parameters
+    ----------
+    sensor_data_sequence : sequence of sensor data objects
+        These objects may include :class:`SensorData`, :class:`numpy.ndarray`,
+        :class:`CategoricalData` and lists of sensor values
+
+    Returns
+    -------
+    dtypes : list of :class:`numpy.dtype` objects, or [None]
+        List of unique dtypes found in sequence, or [None] if dtype is unknown
+
+    """
+    dtypes = [infer_dtype(sd) for sd in sensor_data_sequence]
+    # Ignore unavailable dtypes unless nothing is available
+    dtypes = [dt for dt in dtypes if dt is not None]
+    dtypes = unique_in_order(dtypes) if dtypes else [None]
+    if all(np.issubdtype(dtype, np.str) for dtype in dtypes):
+        # Strings of different lengths have different dtypes
+        # (e.g. '|S1' vs '|S10') but can be safely concatenated
+        dtypes = [np.dtype('|S%d' % (max([dt.itemsize for dt in dtypes]),))]
```

Comment:
The docstring implies that starting with a mix of int and string types, the string types should all be collapsed to the longest string type, but the code doesn't match that. Looking at the usage I suspect that isn't terribly important (you just want to know if there is a unique dtype), so it may be easier to just fix up the docstring.

This function would also be an ideal candidate for a unit test...